### PR TITLE
Add support for Python based models to MLIAP package

### PR DESCRIPTION
**Pull request by athomps**
**Summary**

Added a new mliap model style *mliappy *that provides a general interface to Python-based machine-learning models, including PyTorch.

**Related Issue(s)**

n/a

**Author(s)**

Nicholas Lubbers, LANL
Aidan Thompson, SNL

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

The addition of the *mliappy* model style introduces a dependency on Python that did not exist before.  

**Implementation Notes**

Getting this to work required @lubbersnick adding some code to the LAMMPS Python interface (python/lammps.py,  src/PYTHON/python_impl.cpp, cmake/Modules/Packages/PYTHON.cmake)

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by athomps** [comment]
Building LAMMPS with MLIAP package requires first running these commands:

```
cd src/MLIAP
cythonize mliap_model_python_couple.pyx
```

**Comment by athomps** [comment]
This PR needs work that is beyond my expertise in order to pass the build tests.  Specifically, there needs to be a way to either:
a) run the cythonize command mentioned above before building
or b) make the -DLMP_MLIAPPY flag optional , so that default setting eliminates Python dependence e.g. CMakeLists.txt:if(WITH_MLIAPPY).  Any suggestions appreciated.

**Comment by akohlmey** [comment]
> This PR needs work that is beyond my expertise in order to pass the build tests. Specifically, there needs to be a way to either:
> a) run the cythonize command mentioned above before building
> or b) make the -DLMP_MLIAPPY flag optional , so that default setting eliminates Python dependence e.g. CMakeLists.txt:if(WITH_MLIAPPY). Any suggestions appreciated.

i can give it a try.

**Comment by athomps** [comment]
@akohlmey Thanks!  If you have questions about the Python aspects, @lubbersnick can explain.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (rbberger) [inline cmake/Modules/Packages/MLIAP.cmake:1]: We should look for a way that does not produce files in the source directory `../src/MLIAP`

Comment (akohlmey) [reply cmake/Modules/Packages/MLIAP.cmake:1]: already on it.

Comment (akohlmey) [reply cmake/Modules/Packages/MLIAP.cmake]: just pushed what I currently have to see how it fares with integration testing after it is (mostly) working on my local machine. still need to update documentation.

**Comment by athomps** [comment]
Thanks @akohlmey and @rbberger.  This looks very nice. I noticed with cmake build, if MLIAP_ENABLE_PYTHON on and PKG_PYTHON off, no error is detected during configuration, only when
 compiling mliap_model_python_couple.cpp.

**Comment by akohlmey** [comment]
> Thanks @akohlmey and @rbberger. This looks very nice. I noticed with cmake build, if MLIAP_ENABLE_PYTHON on and PKG_PYTHON off, no error is detected during configuration, only when
> compiling mliap_model_python_couple.cpp.

you probably have a leftover copy in src/MLIAP/ that should be deleted. it will be created on the fly in either src or build/cython during compilation/configuration.

**Comment by akohlmey** [comment]
@athomps please note this was just pushed as far as needed to get the basics working.careful testing is still pending and will happen tomorrow. also, I have not done any checking on the python part of the code. that is something better done by @rbberger. I hope that @junghans can have a look at the CMake code for obvious mistakes and @sjplimp can check if there are any obvious oversights on the conventional build side.

**Comment by athomps** [comment]
@akohlmey Understood. The build instructions in src/MLIAP/README, examples/mliap/README, and doc/src/Packages_details.rst also need to be updated, but I will wait until everything else is resolved.

**Comment by akohlmey** [comment]
> @akohlmey Understood. The build instructions in src/MLIAP/README, examples/mliap/README, and doc/src/Packages_details.rst also need to be updated, but I will wait until everything else is resolved.

Package_details.rst has been updated as needed. The build instructions themselves have been added to Build_extras.rst where they also are for other packages. I didn't update the README files. Feel free to do so. I won't touch it until tomorrow morning.

**Comment by athomps** [comment]
Building LAMMPS with MLIAP package and with/without PYTHON package works as expected.  The cmake dependency checking for `(MLIAP_ENABLE_PYTHON=on && PKG_PYTHON=off)` is not working for me.  It is working for `(PKG_MLIAP=on && PKG_SNAP=off)`

**Comment by athomps** [comment]
> I didn't update the README files. Feel free to do so. I won't touch it until tomorrow morning.

I'll wait until everything is working.

**Comment by akohlmey** [comment]
@rbberger I will tried my best to make it compile and auto adapt to include python support when possible by default.

> Building LAMMPS with `make yes-mliap` and `make yes/no-python` works as expected. The cmake dependency checking for `(MLIAP_ENABLE_PYTHON=on && PKG_PYTHON=off)` is not working for me. It is working for `(PKG_MLIAP=on && PKG_SNAP=off)`

should be fixed now

**Comment by athomps** [comment]
> should be fixed now

Yes, cmake detects the dependence now.

File: cmake/Modules/Packages/PYTHON.cmake
Comment (junghans) [inline cmake/Modules/Packages/PYTHON.cmake:7]: Why is this needed? Should we included in the target below.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [inline cmake/Modules/Packages/MLIAP.cmake:15]: The message reads a bit strange.

File: cmake/Modules/Packages/PYTHON.cmake
Comment (athomps) [reply cmake/Modules/Packages/PYTHON.cmake:7]: I think that might be no longer needed. @lubbersnick can you confirm?

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [inline cmake/Modules/Packages/MLIAP.cmake:21]: Do you really have to copy it? Just refer to it using the full path.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (athomps) [reply cmake/Modules/Packages/MLIAP.cmake:15]: It does the job. I will review all the install instructions at the end and try to harmonize them.

Comment (lubbersnick) [reply cmake/Modules/Packages/MLIAP.cmake:15]: Thought: we might change the run-time behavior to put a message in like this right now. As-is if MLIAP+python support is not enabled, it will just give "Invalid Pair Style".

File: cmake/Modules/Packages/PYTHON.cmake
Comment (lubbersnick) [reply cmake/Modules/Packages/PYTHON.cmake:7]: My build works without it. It was probably introduced accidentally.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (akohlmey) [reply cmake/Modules/Packages/MLIAP.cmake:21]: Yes, this is required. `cythonize` will create the source code at the location of the the .pyx file and we don't want created files mixed with the source.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [reply cmake/Modules/Packages/MLIAP.cmake:15]: I would maybe say "Python package must be enabled for MLIAP with Python support"

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [reply cmake/Modules/Packages/MLIAP.cmake:21]: Thanks,

File: cmake/Modules/FindCythonize.cmake
Comment (junghans) [inline cmake/Modules/FindCythonize.cmake:11]: We are implicitly requiring python-3.6 here, is that wanted?

Comment (akohlmey) [reply cmake/Modules/FindCythonize.cmake:11]: yes

Comment (lubbersnick) [reply cmake/Modules/FindCythonize.cmake:11]: Probably we want python >=3.5, although for certain aspects >=3.6 might be preferable. It is definitely not constrained to exactly 3.6. I have mostly been testing on 3.7. Technically cython should work on 3.3+

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [reply cmake/Modules/Packages/MLIAP.cmake:21]: we could use `copy_if_different` instead of `copy` here.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (junghans) [inline cmake/Modules/Packages/MLIAP.cmake:22]: For readability I would move the `-3` to the front.

File: cmake/Modules/FindCythonize.cmake
Comment (akohlmey) [reply cmake/Modules/FindCythonize.cmake:11]: Somewhere in the comments or docs it said Python > 3.5 was required. So that is what I implemented.
For your reference: Our lower boundary for target distributions we want to support is CentOS 7, which comes with Python 3.6. Ubuntu 16.04 has Python 3.5, but we are not testing on it (anymore) and we don't support CMake on it, as the CMake version is too old. So version Python 3.6 is the *de facto* lower boundary for Python 3 (and 2.7 the lower and upper boundary on Python 2)

Comment (akohlmey) [reply cmake/Modules/FindCythonize.cmake:11]: > Probably we want python >=3.5, although for certain aspects >=3.6 might be preferable. It is definitely not constrained to exactly 3.6. I have mostly been testing on 3.7. Technically cython should work on 3.3+

FYI, the version number specified with `find_package()` is the lower boundary. So we can require Python 3.6 when using CMake to configure (which has its lower bound at version 3.10) and not apply any additional restrictions to what we already have.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (lubbersnick) [reply cmake/Modules/Packages/MLIAP.cmake:22]: The -3 shouldn't need to be there, the pyx file has a language_level declaration at the top.

Comment (akohlmey) [reply cmake/Modules/Packages/MLIAP.cmake:22]: > The -3 shouldn't need to be there, the pyx file has a language_level declaration at the top.

i removed most of those. many didn't seem to have an effect, anyway. i consider it a bad idea to have command line flags embedded in source code unless it cannot be avoided. the flag was added to make things more explicit.

Comment (lubbersnick) [reply cmake/Modules/Packages/MLIAP.cmake:22]: I had the others around for building this as a standalone module, which is a separate endeavor. I don't have a problem with them removed for lammps:master.

But the cython language_level and distutils : c++ flags are pretty key because they actually give information about how the source file should be read by humans as well, I think it is good to leave them there.

File: cmake/Modules/FindCythonize.cmake
Comment (lubbersnick) [reply cmake/Modules/FindCythonize.cmake:11]: Okay, makes sense -- I tried digging through the cmake docs but couldn't figure out if it was an exact or a lower bound.

File: cmake/Modules/Packages/MLIAP.cmake
Comment (akohlmey) [reply cmake/Modules/Packages/MLIAP.cmake:22]: ...and those two settings had been retained in the source.

**Comment by akohlmey** [comment]
docs, readme files, and build scripts have been updated to (hopefully) satisfy all concerns raised in the discussion. handing the PR over to @rbberger to review and test the Python component and pytorch intergration.

**Comment by junghans** [comment]
With this package we will have three different ways to interact with python:

- through `ctypes` for `PKG_PYTHON`
- through `cythonize` for `PKG_MLIAP`
- through `swig` for the generated interface

Are they all compatible with each other?

File: cmake/Modules/FindCythonize.cmake
Comment (junghans) [reply cmake/Modules/FindCythonize.cmake:11]: Actually we only check if python is >=3.6 not cythonize ifself. CMake could still pick up an earlier version as `_python_path` is only used a hints. Is there a required cython version? I only test this PR with cython 0.29.21

Comment (junghans) [reply cmake/Modules/FindCythonize.cmake:11]: We could use https://github.com/espressomd/espresso/blob/python/cmake/FindCython.cmake#L25-L30 to check the version.

**Review by athomps — COMMENTED** [review]
No comment

File: python/lammps.py
Comment (athomps) [reply python/lammps.py:509]: @nicklubbers says he posted a response to above comments earlier today, but it is awaiting moderator approval.  Can he be added to a  pre-approved list?

Comment (akohlmey) [reply python/lammps.py:509]: > @lubbersnick says he posted a response to above comments earlier today, but it is awaiting moderator approval. Can he be added to a pre-approved list?

I am not aware of having any moderation feature enabled for lammps on GitHub.

Comment (lubbersnick) [reply python/lammps.py:509]: Fixed now. It didn't say anything about "moderators", just listed "pending" next to the comments. I guess it had something to do with my leaving a comment on a part of the code nobody had discussed yet. Sorry about that.

Comment (akohlmey) [reply python/lammps.py:509]: > Fixed now. It didn't say anything about "moderators", just listed "pending" next to the comments. I guess it had something to do with my leaving a comment on a part of the code nobody had discussed yet. Sorry about that.

The "pending" status happens when you click on "Start a review" while the behavior you expected would happen with clicking on the "Add single comment" button.

File: examples/mliap/README
Comment (athomps) [reply examples/mliap/README:63]: It was called test_pylibs.py.  I removed it because it is outside the mission scope of the examples folder, which is to test whether LAMMPS is running correctly, as opposed to checking the configuration of the user's runtime environment.   LAMMPS should not get into the business of testing peoples computers for them, except in the context of cmake.  Feel free to augment the language in the README to make it easier to understand.

File: python/lammps.py
Comment (rbberger) [reply python/lammps.py:509]: @lubbersnick I agree that `lammps` should be a package. I had that on my agenda for some time now, but it was always pushed back for some reason (e.g., stable release). I'll use this as an excuse to start that process. I'll create a new PR for this change.

File: examples/mliap/README
Comment (lubbersnick) [reply examples/mliap/README:63]: Ok, we should put this back into the MLIAPPY object.

File: python/lammps.py
Comment (lubbersnick) [reply python/lammps.py:509]: I am happy to try and help refactoring the file if there is a clear method for me to test my changes, feel free to involve me about that. The part where I'm still pretty new is on the build/configuration side.

For this PR are we ok to add `class MLIAPPY()` to the end of lammps.py? Would you prefer if we moved mliappy_pytorch.py to a new directory /python/mliappy/ for the moment, and update the example so instead of copying, it add this to the sys.path?

File: examples/mliap/README
Comment (athomps) [reply examples/mliap/README:63]: Yes, I agree.

File: python/lammps.py
Comment (rbberger) [reply python/lammps.py:509]: @lubbersnick I started with PR #2521. Let me know what you think. For this PR I'm ok having it in lammps.py for now. I only would ask to only create a MLIAPPY instance when accessed, similar to what you see with the optional numpy wrapper. Having that folder and adding it to the sys path sound good for now. It would probably end up in a subpackage, like `python/lammps/mliappy/`

**Comment by rbberger** [comment]
@athomps @lubbersnick I've resolved the merge conflicts after PR #2521 got merged. You can now proceed to making the adjustments we talked about.

**Review by rbberger — CHANGES_REQUESTED** [review]
The change to `setuptools` break out-of-source compilation/installation with CMake. I creates a folder `python/lammps.egg-info` in the source checkout. That can cause problems with multiple parallel builds using the same source folder.

**Comment by athomps** [comment]
@lubbersnick has addressed the Python interface issues. I have updated the example and src README files and verified that all the examples still work.  I did not provide reference output for in.mliap.pytorch.relu1hidden, as the model may get refined later.  Assuming it passes the tests, it is ready to be merged.

**Review by rbberger — APPROVED** [review]

**Pull request merged**
